Automattic/wp-calypso
# Signup: DSS: Add DangerousHtml component to prevent rendering errors

Normally it's possible to use the special `dangerouslySetInnerHTML` prop of a
component to add raw HTML into that component, but the markup is injected as-is,
which means that any invalid markup will risk breaking React's virtual DOM.
(Generally this results in errors like `Invariant Violation: Expected markup to
render 15 nodes, but rendered 10`).

DangerousHtml creates a tag (of the tag name matching `tag`) and uses the web
API's `innerHTML` to inject the HTML into the tag rather than using
`dangerouslySetInnerHTML`. Using JavaScript to set the HTML causes the browser
to fix any invalid markup and prevents it from breaking the renderer.

Fixes #1890 

## Testing

1. Visit the DSS flow, eg: http://calypso.localhost:3000/start/dss/
2. Click on a theme; Signup will move to step 2.
3. Click the "Back" link at the bottom of the page to return to step 1.
4. Verify the DSS page displays correctly (no sign of weird rendering).
5. Click on another theme.
6. Complete the signup flow.
7. Verify that the site is created and has the theme you chose.
